### PR TITLE
3.6 sebind impl

### PR DESCRIPTION
**Pull request by PatriceJiang**
Re: https://github.com/cocos/3d-tasks/issues/12285

### Changelog

* Implementation of `sebind` 

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check // Manual trigger with `@cocos-robot run test cases` afterward

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: native/cocos/bindings/sebind/class.inl [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+
+
+template <typename T>
+class class_ { //NOLINT
+public:
+    using class_type = T;
+    using context_   = intl::context_db_::context_;
+
+    explicit class_(context_ *ctx) : _ctx(ctx) {}
+    explicit class_(const std::string &name);
```
Comment (dumganhar) [inline native/cocos/bindings/sebind/class.inl]: Need to use ccstd::string now.

File: native/cocos/bindings/sebind/class.inl [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+    template <typename R>
+    friend void genericConstructor(const v8::FunctionCallbackInfo<v8::Value> &);
+};
+
+
+// implements
+
+template <typename T>
+class_<T>::class_(const std::string &name) {
+    _ctx            = intl::context_db_::instance()[name];
```
Comment (dumganhar) [inline native/cocos/bindings/sebind/class.inl]: Follow cocos coding style,  `context_db_` needs to be renamed to `ContextDB`

File: native/cocos/bindings/sebind/class.inl [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+    _ctx->parentProto = parentProto;
+}
+
+template <typename T>
+template <typename... ARGS>
+class_<T> &class_<T>::constructor() {
+    using CTYPE = intl::Constructor<intl::TypeList<T, ARGS...>>;
+    using MTYPE = intl::TypeMapping<intl::TypeList<ARGS...>>;
+    static_assert(intl::IsConstructibleWithTypeList<T, typename MTYPE::result_types>::value, "No matched constructor found");
+    auto *constructp      = new CTYPE();
```
Comment (dumganhar) [inline native/cocos/bindings/sebind/class.inl]: Use ccnew instead.

File: .github/workflows/native-unit-test.yml [hunk 4]
```diff
@@ -33,10 +33,94 @@ jobs:
...
           ./tests/unit-test/build/src/Release/CocosTest.exe
-      
+
       #### TODO(PatriceJiang): model split
       # - name: Run Module Tests
       #   shell: bash
       #   run: |
       #     cd tests/module-tests 
       #     bash run-test-windows.sh
+  sebind-with-cmake:
```
Comment (minggo) [inline .github/workflows/native-unit-test.yml]: Use `sebind-test-windows `?

File: native/CMakeLists.txt [hunk 5]
```diff
@@ -27,6 +27,7 @@ cc_set_if_undefined(USE_JOB_SYSTEM_TBB       OFF)
 cc_set_if_undefined(USE_PHYSICS_PHYSX        OFF)
 cc_set_if_undefined(USE_MODULES              OFF)
 cc_set_if_undefined(USE_SERVER_MODE          OFF)
+cc_set_if_undefined(USE_CCACHE               ON)
```
Comment (minggo) [inline native/CMakeLists.txt]: What's `CCACHE`?

File: native/CMakeLists.txt [hunk 6]
```diff
@@ -99,6 +100,13 @@ if(USE_JOB_SYSTEM_TASKFLOW)
     endif()
 endif()
 
+if(ANDROID)
+    if(CCACHE_EXECUTABLE AND USE_CCACHE)
```
Comment (minggo) [inline native/CMakeLists.txt:104]: Where to declare CCACHE_EXECUTABLE?

File: native/CMakeLists.txt [hunk 7]
```diff
@@ -2346,8 +2356,17 @@ elseif(LINUX)
...
+    cocos/bindings/sebind/intl/common.h
+    cocos/bindings/sebind/intl/common.cpp
+    cocos/bindings/sebind/class.inl
+    cocos/bindings/sebind/sebind.h
+)
+
 if(USE_SE_V8)
     cocos_source_files(MODULE ccbindings
+                     cocos/bindings/sebind/class_v8.h
+                     cocos/bindings/sebind/class_v8.cpp
```
Comment (minggo) [inline native/CMakeLists.txt:2370]: File name do not use `_ `.

File: native/CMakeLists.txt (hunk 5 above)
Comment (PatriceJiang) [reply native/CMakeLists.txt]: https://ccache.dev/  A tool to cache & speedup android build predure

File: native/CMakeLists.txt [hunk 8]
```diff
@@ -2346,8 +2356,17 @@ elseif(LINUX)
 endif()
 
 ############# module bindings
+cocos_source_files(MODULE ccbindings
+    cocos/bindings/sebind/intl/common.h
+    cocos/bindings/sebind/intl/common.cpp
+    cocos/bindings/sebind/class.inl
+    cocos/bindings/sebind/sebind.h
```
Comment (minggo) [inline native/CMakeLists.txt:2364]: Use capital character for file name.

File: native/CMakeLists.txt (hunk 6 above)
Comment (PatriceJiang) [reply native/CMakeLists.txt:104]: Set by `find_program` in predefine.cmake

File: native/cocos/bindings/sebind/class.inl [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+
+#include "bindings/manual/jsb_conversions.h"
+#include "bindings/manual/jsb_global.h"
+#include "intl/common.h"
+
+namespace sebind {
+
+
+template <typename T>
+class class_ { //NOLINT
```
Comment (minggo) [inline native/cocos/bindings/sebind/class.inl]: Why use such a class name?

File: native/CMakeLists.txt (hunk 8 above)
Comment (PatriceJiang) [reply native/CMakeLists.txt:2364]: Should this be rename to `SEBind.h` ?

File: native/cocos/bindings/sebind/class_v8.h [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+ THE SOFTWARE.
+****************************************************************************/
+#pragma once
+
+#include "class.inl"
+
+namespace sebind {
+
+template <typename T>
+void genericFinalizer(se::Object *obj) {
```
Comment (minggo) [inline native/cocos/bindings/sebind/class_v8.h:32]: Please add comments for class, functions.

File: native/cocos/bindings/sebind/class.inl (hunk 9 above)
Comment (PatriceJiang) [reply native/cocos/bindings/sebind/class.inl]: These names are borrowed from embind.

File: .github/workflows/native-unit-test.yml (hunk 4 above)
Comment (PatriceJiang) [reply .github/workflows/native-unit-test.yml]: Yep, I will fix this.

File: native/cocos/bindings/sebind/class_v8.h (hunk 10 above)
Comment (PatriceJiang) [reply native/cocos/bindings/sebind/class_v8.h:32]: OK

File: native/CMakeLists.txt (hunk 5 above)
Comment (minggo) [reply native/CMakeLists.txt]: Any bad effect?

Comment (PatriceJiang) [reply native/CMakeLists.txt]: Perhaps disk usage

File: native/cocos/bindings/sebind/intl/common.h [hunk 11]
```diff
@@ -0,0 +1,1003 @@
...
+    using type = void (*)(T *);
+    using arg_type = T;
+    type func;
+    void finalize(void *ptr) override {
+        (*func)(reinterpret_cast<T *>(ptr));
+    }
+};
+
+template <typename T, typename R, typename... ARGS>
+struct InstanceMethod<R (T::*)(ARGS...)> : InstanceMethodBase {
```
Comment (dumganhar) [inline native/cocos/bindings/sebind/intl/common.h:514]: Is it better to use public inheritance?
```c++
template <typename T, typename R, typename... ARGS>
struct InstanceMethod<R (T::*)(ARGS...)> : public InstanceMethodBase {
```

Then if we get a `InstanceMethodBase` pointer, we could invoke its virtual functions.

File: native/CMakeLists.txt (hunk 5 above)
Comment (minggo) [reply native/CMakeLists.txt]: Does it cause any cache error?

File: native/cocos/bindings/sebind/intl/common.h (hunk 11 above)
Comment (PatriceJiang) [reply native/cocos/bindings/sebind/intl/common.h:514]: struct inheritance is public by default, the `public` keyword is omitted here.

File: native/cocos/bindings/sebind/class.inl [hunk 12]
```diff
@@ -0,0 +1,404 @@
...
+#include "bindings/manual/jsb_conversions.h"
+#include "bindings/manual/jsb_global.h"
+#include "intl/common.h"
+
+namespace sebind {
+
+/**
+ * @brief Export C++ Class/Funtions to JS
+ * 
+ * @tparam T  C++ type
```
Comment (minggo) [inline native/cocos/bindings/sebind/class.inl:50]: `tparam` typo?

File: native/cocos/bindings/sebind/class.inl [hunk 13]
```diff
@@ -0,0 +1,404 @@
...
+public:
+    using class_type = T;
+    using Context   = intl::ContextDB::Context;
+
+    explicit class_(Context *ctx) : _ctx(ctx) {}
+
+    /**
+     * @brief Construct a new class_ object
+     * 
+     * @param name specify the class in JS
```
Comment (minggo) [inline native/cocos/bindings/sebind/class.inl]: class -> class name?

File: native/cocos/bindings/sebind/class.inl [hunk 14]
```diff
@@ -0,0 +1,404 @@
...
+    template <typename... ARGS>
+    class_ &constructor();
+
+    /**
+     * @brief Define a constructor by a function
+     * The signature of the function pointer can be
+     * - `bool(*)(se::State&)`
+     * or
+     * - `T*(*)(ARGS...)`
+     * @tparam F
```
Comment (minggo) [inline native/cocos/bindings/sebind/class.inl:108]: `@tparam`, typo?

File: native/cocos/bindings/sebind/class.inl [hunk 15]
```diff
@@ -0,0 +1,404 @@
...
+     * @param method Member function pointer of class `T`, or normal function which the first argument is `T*`.
+     * @return class_& 
+     */
+    template <typename Method>
+    class_ &function(const char *name, Method method);
+
+    /**
+     * @brief Define a property for js class
+     * 
+     * @tparam Field
```
Comment (minggo) [inline native/cocos/bindings/sebind/class.inl:136]: The same as above.

File: native/tests/sebind-tests/data/main.js [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+
+    demo.Coconut.prototype.rock = function () {
+        return "Coconut Rock";
+    }
+    let c = new demo.Coconut();
+    block("call js", () => {
+        assert(c.callJsFunction("rock") == "Coconut Rock", "Call JS function");
+    });
+    block("call setRadius", () => {
+        assert(typeof c.setRadius === 'function', "functio setRadius");
```
Comment (minggo) [inline native/tests/sebind-tests/data/main.js]: `functio` typo.

File: native/tests/sebind-tests/data/main.js [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+    });
+    block("call getRadius", () => {
+        assert(typeof c.getRadius === 'function', "functio getRadius");
```
Comment (minggo) [inline native/tests/sebind-tests/data/main.js]: the same as above

File: native/tests/sebind-tests/data/main.js [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+    });
+    block("call static 123", () => {
+        assert(typeof demo.Coconut.OneTwoThree === 'function', "static function getRadius");
```
Comment (minggo) [inline native/tests/sebind-tests/data/main.js]: the same as above

File: native/tests/sebind-tests/data/main.js [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+    let d = new demo.CoconutExt;
+    block("check time", () => {
+        l(`time is ${demo.CoconutExt.time}`)
+    });
+    block("check getTime", () => {
+        assert(typeof demo.CoconutExt.getTime === 'function', "static time");
+        l(`gettime is ${demo.CoconutExt.getTime()}`);
+    });
+    block("check area", () => {
+        l(`area is ${d.area}`);
```
Comment (minggo) [inline native/tests/sebind-tests/data/main.js:92]: I think log is not good to use in unit test, as can not check the result in code.

Comment (PatriceJiang) [reply native/tests/sebind-tests/data/main.js:92]: Yes, assertion logic is added.

File: native/cocos/bindings/manual/jsb_conversions.h [hunk 20]
```diff
@@ -635,9 +635,9 @@ struct is_variant<cc::variant<Args...>> : std::true_type {}; // NOLINT
 template <typename T>
 inline typename std::enable_if_t<!std::is_enum<T>::value && !std::is_pointer<T>::value && !is_jsb_object_v<T>, bool>
 sevalue_to_native(const se::Value & /*from*/, T * /*to*/, se::Object * /*unused*/) { // NOLINT(readability-identifier-naming)
-    SE_LOGE("Can not convert type ???\n - [[ %s ]]\n", typeid(T).name());
+    SE_LOGE("Missing convertion for type [[%s]]\n", typeid(T).name());
```
Comment (PatriceJiang) [inline native/cocos/bindings/manual/jsb_conversions.h]:
```suggestion
    SE_LOGE("Missing conversion impl `sevalue_to_native` for type [[%s]]\n", typeid(T).name());
```

File: native/cocos/bindings/sebind/class.inl [hunk 21]
```diff
@@ -0,0 +1,410 @@
...
+
+namespace sebind {
+
+/**
+ * @brief Export C++ Class/Funtions to JS
+ * 
+ * @tparam T  C++ type
+ */
+template <typename T>
+class class_ { //NOLINT
```
Comment (dumganhar) [inline native/cocos/bindings/sebind/class.inl]: Add `final` keyword for classes that should not be inherited.

File: native/cocos/bindings/sebind/class.inl [hunk 22]
```diff
@@ -0,0 +1,410 @@
...
+     */
+    explicit class_(const char *name, se::Object *parentProto);
+
+    ~class_() {
+        assert(_installed); // procedure `class_::install` has not been invoked?
+    }
+
+    /**
+     * @brief  Attach the function object to the namespace object, 
+     * then the export class can be accessed through the namespace object in JS.
```
Comment (dumganhar) [inline native/cocos/bindings/sebind/class.inl]: the `exported` class

File: native/cocos/bindings/sebind/class.inl (hunk 13 above)
Comment (PatriceJiang) [reply native/cocos/bindings/sebind/class.inl]: fixed

File: native/tests/sebind-tests/data/main.js (hunk 16 above)
Comment (PatriceJiang) [reply native/tests/sebind-tests/data/main.js]: fixed

File: native/tests/sebind-tests/data/main.js (hunk 17 above)
Comment (PatriceJiang) [reply native/tests/sebind-tests/data/main.js]: fixed

File: native/tests/sebind-tests/data/main.js (hunk 18 above)
Comment (PatriceJiang) [reply native/tests/sebind-tests/data/main.js]: fixed

File: native/cocos/bindings/sebind/class.inl (hunk 21 above)
Comment (PatriceJiang) [reply native/cocos/bindings/sebind/class.inl]: fixed

File: native/cocos/bindings/sebind/class.inl (hunk 22 above)
Comment (PatriceJiang) [reply native/cocos/bindings/sebind/class.inl]: fixed

File: native/CMakeLists.txt (hunk 5 above)
Comment (PatriceJiang) [reply native/CMakeLists.txt]: `ccache` is a mature tool supported by Android. The default configuration is changed to off, which can be enabled by developers themself. We need to add an option on the build panel.

File: native/cocos/bindings/sebind/class.inl [hunk 23]
```diff
@@ -0,0 +1,418 @@
...
+
+template <typename T>
+template <typename Getter, typename Setter>
+class_<T> &class_<T>::property(const char *name, Getter getter, Setter setter) {
+    using ATYPE       = intl::InstanceAttribute<intl::AttributeAccessor<T, Getter, Setter>>;
+    auto *attrp       = ccnew ATYPE();
+    attrp->getterPtr  = ATYPE::HAS_GETTER ? getter : nullptr;
+    attrp->setterPtr  = ATYPE::HAS_SETTER ? setter : nullptr;
+    attrp->class_name = _ctx->className;
+    attrp->attr_name  = name;
```
Comment (dumganhar) [inline native/cocos/bindings/sebind/class.inl]: Code format:

attrp->className
attrp->attrName

File: native/cocos/bindings/sebind/class.inl [hunk 24]
```diff
@@ -0,0 +1,418 @@
...
+
+template <typename T>
+template <typename Method>
+class_<T> &class_<T>::staticFunction(const char *name, Method method) {
+    using MTYPE          = intl::StaticMethod<Method>;
+    auto *methodp        = ccnew MTYPE();
+    methodp->method_name = name;
+    methodp->class_name  = _ctx->className;
+    methodp->arg_count   = MTYPE::ARG_N;
+    methodp->func        = method;
```
Comment (dumganhar) [inline native/cocos/bindings/sebind/class.inl]: Code format:

methodp->methodName
methodp->className
methodp->argCount

File: native/cocos/bindings/sebind/class_v8.h [hunk 25]
```diff
@@ -0,0 +1,194 @@
...
+        _ctx->kls->defineProperty(std::get<0>(attr).c_str(), getter, setter, std::get<1>(attr).get());
+    }
+    auto *fieldGetter = &genericAccessorGet<intl::InstanceFieldBase>;
+    auto *fieldSetter = &genericAccessorSet<intl::InstanceFieldBase>;
+    for (auto &field : _ctx->fields) {
+        _ctx->kls->defineProperty(std::get<0>(field).c_str(), fieldGetter, fieldSetter, std::get<1>(field).get());
+    }
+    // defineFunctions
+    {
+        std::map<std::string, std::vector<intl::InstanceMethodBase *>> multimap;
```
Comment (dumganhar) [inline native/cocos/bindings/sebind/class_v8.h]: Use ccstd::map, ccstd::vector instead.

File: native/cocos/bindings/sebind/intl/common.h [hunk 26]
```diff
@@ -0,0 +1,1003 @@
...
+struct FunctionWrapper;
+
+template <typename R, typename C, typename... ARGS>
+struct FunctionWrapper<R (*)(C *, ARGS...)> {
+    using type = R (*)(C *, ARGS...);
+    using return_type = R;
+    using arg_list = TypeList<ARGS...>;
+    static constexpr size_t ARG_N = sizeof...(ARGS);
+    template <typename... ARGS2>
+    inline R static invoke(type func, C *self, ARGS2 &&... args) {
```
Comment (dumganhar) [inline native/cocos/bindings/sebind/intl/common.h]: Put `R` after static

File: native/cocos/bindings/sebind/intl/common.h [hunk 27]
```diff
@@ -0,0 +1,1003 @@
...
+
+template <>
+struct FunctionWrapper<std::nullptr_t> {
+    using type = std::nullptr_t;
+    using return_type = std::nullptr_t;
+    ;
+    using arg_list = TypeList<>;
+    static constexpr size_t ARG_N = 0;
+    template <typename C, typename... ARGS>
+    void static invoke(type /*func*/, C * /*self*/, ARGS &&... /*args*/) {
```
Comment (dumganhar) [inline native/cocos/bindings/sebind/intl/common.h]: void static -> static void

File: native/cocos/bindings/sebind/intl/common.h [hunk 28]
```diff
@@ -0,0 +1,1003 @@
...
+    inline R static invoke(type func, C *self, ARGS2 &&... args) {
+        return (self->*func)(std::forward<ARGS2>(args)...);
+    }
+};
+
+template <>
+struct FunctionWrapper<std::nullptr_t> {
+    using type = std::nullptr_t;
+    using return_type = std::nullptr_t;
+    ;
```
Comment (dumganhar) [inline native/cocos/bindings/sebind/intl/common.h]: unused `;` ?

File: native/cocos/bindings/sebind/intl/common.h [hunk 29]
```diff
@@ -0,0 +1,1003 @@
...
+    }
+    virtual bool set(se::State &state) const {
+        if (bfnSetPtr) return (*bfnSetPtr)(state);
+        return false;
+    }
+};
+
+struct StaticMethodBase {
+    std::string class_name;
+    std::string method_name;
```
Comment (dumganhar) [inline native/cocos/bindings/sebind/intl/common.h]: Use ccstd::string instead.

File: native/cocos/bindings/sebind/intl/common.h [hunk 30]
```diff
@@ -0,0 +1,1003 @@
...
+    size_t arg_count;
```
Comment (dumganhar) [inline native/cocos/bindings/sebind/intl/common.h]: Set default value to 0

**Review by dumganhar — APPROVED** [review]

**Review by minggo — APPROVED** [review]

**Comment by minggo** [comment]
Please fix conflict.

**Pull request merged**